### DISCUSSION TITLE
Inserting a drop down based on linked tables

### DISCUSSION THREAD

**Discussion by craigworth** [Budibase/budibase#3655] [Help]
Apologies if this has been asked before. 

I am trying to get a form completed where a user puts in a valuation and a date against a fund. Using a MySQL Database. I have created a link between the valuation table and the fund table using the foreign key which exists in the database. The fund table contains the ID and name of the fund. 

What I would like to do is to use that relationship so the user selects the name of the fund from a drop down and the form then add a line to the valuations table. The valuations table will only have the ID of the fund and not the name. That only exists in the Funds table (linked by the foreign key). 

I just feel I am missing something obvious where I either need to create a new form from scratch and put in the SQL myself to join the tables or the canned option of creating direct from the table should have picked up that link.

Any help appreciated.

**Comment by mshiner** [comment]
I was going to ask the exact same question.

Does anybody out there have a solution to this?

Thanks
M

**Comment by melohagan** [comment]
If I understand this correctly, this may be of use: https://docs.budibase.com/docs/formula#fields-from-relationships

**Reply by mshiner** [reply]
Hmmm, thanks for the link . Not sure.

Example.... given two tables (from our live db but simplified)

**1. Departments**
```
CREATE TABLE IF NOT EXISTS public.mf_depts
(
    id bigint NOT NULL DEFAULT nextval('mf_depts_id_seq'::regclass),
    dept_code character varying COLLATE pg_catalog."default" NOT NULL,
    dept character varying COLLATE pg_catalog."default" NOT NULL,
    production_recording boolean,
    CONSTRAINT mf_depts_pkey PRIMARY KEY (id),
    CONSTRAINT mf_depts_dept_code_key UNIQUE (dept_code)
)
```
**2. Work centres**
```
CREATE TABLE IF NOT EXISTS public.mf_centres
(
    id bigint NOT NULL DEFAULT nextval('mf_centres_id_seq'::regclass),
    work_centre character varying COLLATE pg_catalog."default" NOT NULL,
    centre character varying COLLATE pg_catalog."default" NOT NULL,
    mfdept_id integer NOT NULL,
    centre_rate double precision NOT NULL,
    production_recording boolean,
    available_qty integer NOT NULL DEFAULT 1,
    CONSTRAINT mf_centres_pkey PRIMARY KEY (id),
    CONSTRAINT mf_centres_centre_code_key UNIQUE (work_centre),
    CONSTRAINT mf_centres_mfdept_id_fkey FOREIGN KEY (mfdept_id)
        REFERENCES public.mf_depts (id) MATCH SIMPLE
        ON UPDATE CASCADE
        ON DELETE NO ACTION
)
```
There is a one-to-many FK relationship between departments and work centres. If I'm adding a workcentre I need a field on the form for mfdept_id that actually shows the availble departments in, say, a drop down or selector. From a UI persective I don't want to display the list of department ids, I need to display the dept_code column or better still a concatenation of dept_code and department. When I save the form though I need the departmnet id saved in the work centre table.

In our php MVC framework we'd do something in the models like:

1. Departments model
```
// Set specific characteristics
		$this->idField='id';
		$this->orderby='dept_code';
		$this->identifierField='dept_code||\' - \'||dept';
 		$this->validateUniquenessOf('dept_code'); 
 		
// Define relationships
 		$this->hasMany('MFCentre','centres','mfdept_id');
```
2. Work Centres model
```
// Set specific characteristics
		$this->idField='id';
		$this->orderby='work_centre';
		$this->identifierField='work_centre || \' - \' ||centre';
 		$this->validateUniquenessOf('work_centre');
 		
// Define relationships
 		$this->belongsTo('MFDept', 'mfdept_id', 'mfdept'); 
 		$this->hasOne('MFDept', 'mfdept_id', 'dept_detail');
```
And then use a smarty template and some JS (jquery) to build the form in the controller view/edit function. Its a bit 'old school' but works. 

Problem is for someone who isn't steeped in the 'uzERP way' its hard yards to add new stuff while utilising all the data that is already there.

Regards
M

PS - we are going to have to change those FK referencing field types to match the id type in the referenced table to make budibase work with the uzERP tables..... unless you can do the casting as in the other thread.

**Comment by melohagan** [comment]
Hey @mshiner 

Hopefully this will help. 
As mentioned, the data types of `mf_depts.id` and `mf_centres.mfdept_id` would need to match for this to work.

### Displaying the department dropdown
1. Add a formula to the *mf_depts* table for displaying the code and dept together:
    ![Screenshot 2022-05-18 at 10 20 01](https://user-images.githubusercontent.com/101575380/169004993-82dc7b42-d48a-4c7e-8425-d869c6cce17f.png)
    ![Screenshot 2022-05-18 at 10 20 19](https://user-images.githubusercontent.com/101575380/169005051-ec5a6780-d82d-45d2-932a-53ccd6e0811b.png)

2. In the 'create new work centre' screen, add a Data Provider for your *mf_depts* table. The form with the *mf_centres* schema should be nested underneath.  Add an option picker to the form and type in a unique **field name**.
![Screenshot 2022-05-18 at 10 59 30](https://user-images.githubusercontent.com/101575380/169013326-4984e97f-dddc-44a5-8223-bf4e283843aa.png)

3. Select the data provider as the options source. 
Also set the label column to the formula column, and the value to the id. 
![Screenshot 2022-05-18 at 11 01 47](https://user-images.githubusercontent.com/101575380/169013764-3426021e-afd3-4b0b-aec8-24a8ab3f7219.png)
 
### Save the chosen department against the new work centre
1. Define actions for the save button. This should save a new row to the *mf_centres* table from the form as normal, but we also want to 'Add Column' so that we can save the chosen **id** from the dropdown to the relationship column:
![Screenshot 2022-05-18 at 11 03 12](https://user-images.githubusercontent.com/101575380/169014013-5ad5eae9-8a2b-406e-8ad5-07f2d0b6ee87.png)
The binding in this case is: `{{ New Work Centre Form.Fields.Department }}`

----
Just as a note, if you want to be able to save to multiple tables in a single form, there's a couple of examples here:
- https://docs.budibase.com/docs/form-steps
- https://docs.budibase.com/docs/referencing#example-creating-and-linking-two-table-entries-in-one-form

**Reply by mshiner** [reply]
**BRILLIANT!!!!** - that gets me up the hill very quickly - essentially I can now pretty much define my 'models' in Budibase and start the process of adding some front end functionality to uzERP.

M

**Comment by mshiner** [comment]
Hey @melohagan 

This solution works well (obviously).

I have managed to set up some FK relationships correctly (by making sure the column type matches). I have also set a 'caluculated column' as the table display column. Example:

Person table has column id, Firstname and Surname in postgreSQL
Set up a new column Full Name and set as ttable display
In hours worked table there is a FK column person_id 
Set up relationship person.id to hours.person_id

The column 'automagically' shows the full name in the hours view.... cool

I have then autogenerated CRUD screens and a relationship picker in the hours view appears with the full name as the option to pick - GREAT!

But when I try to add a record the app complains about the FK column not being set - I looked at the relationship picker and there is no way to specifiy what to save (i.e. the id) which I guess is the problem. 

Is there something else I need to do?

M

**Reply by melohagan** [reply]
In the field dropdown of your relationship picker, are you able to select the FK?
![Screenshot 2022-06-01 at 10 45 23](https://user-images.githubusercontent.com/101575380/171376325-af4e72f9-ba27-453d-ab69-f6b37ff8a982.png)

**Reply by mshiner** [reply]
@melohagan 

Yep - so in my case its hours.person (i.e. the name I gave it when setting the relationship). I need the id saved in the hours table.

I also tried it with a 'Nullable' FK (i.e. where the column in the hours table can be null but if NOT NULL needs to be from another table). In this case it is a table called hour_types with an id and description. I tried it with both the id and the description columns as the display column on this table - on saving the record in the hours table the hour_type_id (FK id of hour_types) is not saved in table - note I forced the person to check this wasn't the error.

M

**Reply by mshiner** [reply]
Ooops - forgot to say - this is postgreSQL and the column types all match.

M

**Reply by melohagan** [reply]
@mshiner I recreated something similar in postgres and I was able to save a new **mf_dept** row after selecting an **mf_centre** in the relationship picker.

Would you mind sharing your app export to support@budibase.com and I can have a look.

**Reply by mshiner** [reply]
@melohagan 

Will do

M

**Reply by melohagan** [reply]
So I haven't tested this out, but I think it may work if you remove the **person_id** override. 

The schema will pick up your **hours.person** field, and should automatically assign the **person_id** field a value. Assuming the foreign key relationship has been setup correctly.

![Screenshot 2022-06-01 at 14 46 41](https://user-images.githubusercontent.com/101575380/171419834-3d64fa29-d514-4cf4-82ce-0edd1fefe69c.png)

**Reply by mshiner** [reply]
Hi @melohagan 

I tried this - originally I didn't have the picker set up but was just trying to use the relationship picker to achieve the same thing as you scoped out above (with a datasource based standard picker). This didn't work so I used your method.

When I added the type_id field as a relationship I automagically got a relationship picker - cool. But it doesn't save the id in the column on the hours table.
